Enable/test torch.arange for Q2 target models

### What feature would you like to be added?

Enable torch.arange for Q2 target models

### Why is this feature needed?

Mistral-Small-3-2-24B-Instruct-2506, granite-3-3-8b-instruct-fms, and granite-4-0-h-tiny need torch.arange

### Any alternative solution that you may want to share?

_No response_

### Anything else we need to know?

While it was originally enabled at https://github.com/torch-spyre/torch-spyre/issues/120, a new test case from Mistral fails

The torch.arange is decomposed to torch.ops.prims.iota.default, which places an output tensor on CPU, for a use case from Mistral while the output tensor is supposed to on Spyre. For other use cases, it is directly decomposed to torch.ops.aten.arange.default for other use cases which places the output tensor on Spyre while the operation is off-loaded to CPU.

<details>
<summary>Bob's report</summary>

# torch.arange Operation Test Failure Analysis

**Analysis Date:** 2026-04-07  
**Test Run Date:** 2026-04-06 08:35:38  
**Operator:** `torch.arange`

---

## Executive Summary

Analysis of `torch.arange` operation test failures reveals **1 failed test case** out of 11 total tests (excluding gpt-oss-20b results as requested). The failure is specific to the Mistral-Small-3.2-24B-Instruct-2506 model and is caused by **incorrect device placement** where the operation is being executed on CPU instead of the Spyre device.

### Key Findings

- **Root Cause:** torch.arange with int64 dtype is being decomposed to `torch.ops.prims.iota.default` with explicit `device='cpu'` placement
- **Failure Pattern:** Device mismatch - output tensor is on CPU instead of Spyre
- **Data Type:** int64 (unsupported natively on Spyre for certain operations)
- **Impact:** Affects operations that generate integer sequences for indexing/meshing operations

---

## Test Results Overview

### Summary Statistics (Excluding gpt-oss-20b)

| Metric | Count |
|--------|-------|
| **Total Tests** | 10 |
| **Passed** | 9 |
| **Failed** | 1 |
| **Success Rate** | 90% |

### Test Case Breakdown

| Test Case | Model | Status | Device | Dtype |
|-----------|-------|--------|--------|-------|
| torch_arange_1_spyre | Mistral-Small-3.2-24B-Instruct-2506 | ❌ FAILED | spyre | float16 |
| torch_arange_1_spyre | granite-3.3-8b-instruct-fms | ✅ PASSED | spyre | float16 |
| torch_arange_1_spyre | granite-4.0-h-tiny | ✅ PASSED | spyre | float16 |
| torch_arange_2_spyre | granite-4.0-h-tiny | ✅ PASSED | spyre | float16 |
| torch_arange_3_spyre | granite-4.0-h-tiny | ✅ PASSED | spyre | float16 |
| torch_arange_4_spyre | granite-4.0-h-tiny | ✅ PASSED | spyre | float16 |
| torch_arange_5_spyre | granite-4.0-h-tiny | ✅ PASSED | spyre | float16 |
| torch_arange_6_spyre | granite-4.0-h-tiny | ✅ PASSED | spyre | float16 |
| torch_arange_7_spyre | granite-4.0-h-tiny | ✅ PASSED | spyre | float16 |
| torch_arange_8_spyre | granite-4.0-h-tiny | ✅ PASSED | spyre | float16 |

---

## Detailed Failure Analysis

### Failed Test Case: Mistral-Small-3.2-24B-Instruct-2506

#### Test Information

- **Test Name:** `test_model_ops_db_torch_arange_1_spyre__Mistral-Small-3_2-24B-Instruct-2506_spyre_yaml__spyre_float16`
- **Model:** Mistral-Small-3.2-24B-Instruct-2506
- **Source Location:** `site-packages/transformers/models/pixtral/modeling_pixtral.py:41`
- **Use Case:** Position IDs in meshgrid generation
- **Code Context:**
  ```python
  mesh = torch.meshgrid(torch.arange(height), torch.arange(width), indexing="ij")
  ```

#### Error Details

```
AssertionError: this result must be on spyre
```

**Error Location:** `tests/models/runner.py:383`

The test expects the output tensor to be on the Spyre device, but it's actually on CPU.

#### Input Parameters

- **Input:** `28` (scalar integer - the end value for arange)
- **Expected Output Shape:** `[28]`
- **Expected Output Dtype:** `int64`
- **Expected Device:** `spyre:0`
- **Actual Device:** `cpu`

---

## Compiler Artifacts Analysis

### FX Graph Analysis

#### Failed Case - FX Graph (Readable)

```python
class <lambda>(torch.nn.Module):
    def forward(self):
        # Decomposed to prims.iota with explicit CPU device
        iota: "i64[28]" = torch.ops.prims.iota.default(
            28, 
            start=0, 
            step=1, 
            dtype=torch.int64, 
            device=device(type='cpu'),  # ❌ WRONG DEVICE
            requires_grad=False
        )
        return (iota,)
```

**Key Issue:** The `torch.arange` operation is being decomposed to `torch.ops.prims.iota.default` with an **explicit `device='cpu'` parameter**, forcing execution on CPU instead of Spyre.

#### Passing Case - FX Graph (Readable)

```python
class <lambda>(torch.nn.Module):
    def forward(self):
        # Direct aten.arange call with correct device
        arange: "i64[1]" = torch.ops.aten.arange.default(
            1, 
            device=device(type='spyre'),  # ✅ CORRECT DEVICE
            pin_memory=False
        )
        return (arange,)
```

**Key Difference:** Passing tests use `torch.ops.aten.arange.default` directly with the correct Spyre device specification.

### IR Analysis

#### Failed Case - IR Pre-Fusion

```
op0: SchedulerNode(ComputedBuffer)
op0.writes = [MemoryDep('buf0', c0, {c0: 28})]
op0.outputs = [
    buf0: ComputedBuffer
    buf0.layout = FixedLayout('cpu', torch.int64, size=[28], stride=[1])
                              ^^^^^ CPU device hardcoded
]
op0.group.device = cpu  # ❌ Wrong device
```

The IR clearly shows the buffer is allocated on CPU with `FixedLayout('cpu', torch.int64, ...)`.

#### Failed Case - Generated Code

```python
def call(self, args):
    buf0 = empty_strided_cpu((28, ), (1, ), torch.int64)  # ❌ CPU allocation
    cpp_fused_arange_0(buf0)
    return (buf0, )
```

The generated code uses `empty_strided_cpu` to allocate the buffer on CPU.

#### Passing Case - Generated Code

```python
def call(self, args):
    # Direct call to aten.arange with Spyre device
    buf0 = torch.ops.aten.arange.default(
        1, 
        device=device(type='spyre'),  # ✅ Correct device
        pin_memory=False
    )
    buf1 = buf0
    assert_size_stride(buf1, (1, ), (1, ), 'torch.ops.aten.arange.default')
    assert_alignment(buf1, 16, 'torch.ops.aten.arange.default')
    return (buf1, )
```

Passing tests directly call `torch.ops.aten.arange.default` with the Spyre device, avoiding decomposition.

---

## Root Cause Analysis

### Primary Issue: Call Signature-Dependent Decomposition

The failure occurs due to **different handling of torch.arange call signatures**:

#### Failed Case: Single-Argument Form
```python
torch.arange(28)  # Only 'end' parameter
```
**Compilation Path:**
1. PyTorch decomposes to `torch.ops.prims.iota.default(28, start=0, step=1, dtype=torch.int64, device='cpu', ...)`
2. Device is **hardcoded to 'cpu'** during decomposition
3. Original device specification (`spyre:0`) is **lost**
4. Result: Tensor created on CPU instead of Spyre

#### Passing Cases: Two-Argument Form
```python
torch.arange(0, 64)  # 'start' and 'end' parameters
torch.arange(1)      # Single small value
```
**Compilation Path:**
1. Maps directly to `torch.ops.aten.arange.start` or `torch.ops.aten.arange.default`
2. Device parameter is **preserved** in the aten call
3. No decomposition to prims.iota
4. Result: Tensor correctly created on Spyre

### Decomposition Condition Analysis

Based on the test cases, the decomposition to `prims.iota` occurs when:

| Condition | torch.arange Call | Result | Decomposed? |
|-----------|-------------------|--------|-------------|
| **Single arg, no explicit start** | `torch.arange(28)` | ❌ FAIL | ✅ Yes → prims.iota (CPU) |
| **Two args (start, end)** | `torch.arange(0, 64)` | ✅ PASS | ❌ No → aten.arange.start |
| **Single small arg** | `torch.arange(1)` | ✅ PASS | ❌ No → aten.arange.default |
| **Single medium arg** | `torch.arange(12-18)` | ✅ PASS | ❌ No → aten.arange.default |

**Key Insight:** The decomposition appears to be triggered by:
1. **Single-argument form** with implicit `start=0`
2. **Specific tensor size** (28 in the failed case)
3. **Compilation context** (possibly related to how the operation is traced)

The exact heuristic that triggers decomposition vs. direct aten call is not immediately clear from the artifacts, but the pattern is consistent: **single-argument torch.arange(end) with certain sizes gets decomposed to prims.iota with hardcoded CPU device**.

### Data Type Consideration

The operation uses **int64 dtype**, which is known to have limited support on Spyre:

- **Known Issue:** [Issue #1153](https://github.com/torch-spyre/torch-spyre/issues/1153) - Data type conversion issues
- **Spyre Limitation:** int64 is not natively supported for certain operations on the Spyre AI chip
- **Workaround Needed:** Operations may need to be executed on CPU or converted to supported types

---

## Tensor Shape and Layout Analysis

### Failed Case

| Property | Value |
|----------|-------|
| **Shape** | `[28]` |
| **Dtype** | `int64` |
| **Layout** | Contiguous, stride `[1]` |
| **Device** | `cpu` (incorrect) |
| **Expected Device** | `spyre:0` |
| **Memory Size** | 224 bytes (28 × 8 bytes) |

### Passing Cases

| Test | Shape | Dtype | Device | Status |
|------|-------|-------|--------|--------|
| granite-4.0-h-tiny #1 | `[1]` | `int64` | `spyre` | ✅ Pass |
| granite-4.0-h-tiny #2 | `[1]` | `int64` | `spyre` | ✅ Pass |
| granite-4.0-h-tiny #3-8 | Various | `int64` | `spyre` | ✅ Pass |
| granite-3.3-8b-instruct-fms | `[1]` | `int64` | `spyre` | ✅ Pass |

**Pattern:** Smaller tensor sizes consistently pass, suggesting the decomposition path may be triggered by tensor size or other heuristics.

---

## Related Known Issues

### 1. Unsupported Data Types ([Issue #1153](https://github.com/torch-spyre/torch-spyre/issues/1153))

- **Description:** Certain data type conversions (int64 to float32, float16 to float32) are not supported
- **Relevance:** The failed test uses int64, which has limited Spyre support
- **Impact:** May require CPU fallback or type conversion

### 2. Decomposition Device Handling

- **Issue:** Decomposition passes don't properly propagate device information
- **Manifestation:** prims.iota gets hardcoded `device='cpu'` parameter
- **Solution Needed:** Fix decomposition logic to preserve original device specification

---

## Comparison: Failed vs Passing Tests

### Key Differences

| Aspect | Failed Test | Passing Tests |
|--------|-------------|---------------|
| **Call Signature** | `torch.arange(28)` | `torch.arange(0, 64)` or `torch.arange(1)` |
| **Input Args** | Single arg (end only) | Two args (start, end) or single small value |
| **Tensor Size** | `[28]` | `[1]`, `[12]`, `[13]`, `[64]` |
| **FX Graph Op** | `prims.iota.default` | `aten.arange.start` or `aten.arange.default` |
| **Device in FX** | `device='cpu'` (hardcoded) | `device='spyre'` (preserved) |
| **Code Generation** | CPU kernel (vectorized) | Direct aten call or Spyre fallback |
| **Decomposition** | Yes (to prims.iota) | No (direct aten) |

### Why the Difference?

The key differentiator is the **torch.arange call signature**:

1. **Single-Argument Form:** `torch.arange(end)`
   - Triggers decomposition to `prims.iota` in PyTorch's compilation pipeline
   - Decomposition logic hardcodes `device='cpu'`
   - Loses original device specification
   
2. **Two-Argument Form:** `torch.arange(start, end)`
   - Maps directly to `aten.arange.start` without decomposition
   - Preserves device parameter through compilation
   - Works correctly on Spyre

3. **Small Single Values:** `torch.arange(1)`, `torch.arange(12)`, etc.
   - Also use direct `aten.arange.default` without decomposition
   - May have different optimization path due to size or context
   - Device is preserved

**Note:** The exact heuristic that determines when decomposition occurs is not fully clear from the artifacts alone, but the pattern is consistent across all test cases examined.

---

## Impact Assessment

### Severity: **MEDIUM**

- **Functional Impact:** Operation fails completely for affected patterns
- **Scope:** Limited to specific tensor sizes and decomposition paths
- **Workaround Available:** Yes (see recommendations)

### Affected Use Cases

1. **Position Encoding:** Meshgrid generation for vision transformers (Pixtral model)
2. **Index Generation:** Creating integer sequences for indexing operations
3. **Coordinate Systems:** Generating coordinate grids for spatial operations

### Models Affected

- **Confirmed:** Mistral-Small-3.2-24B-Instruct-2506 (Pixtral vision model)
- **Potentially:** Other models using similar position encoding schemes

---

## Recommendations

### Immediate Fixes

#### 1. Fix Decomposition Device Propagation (Priority: HIGH)

**Location:** Decomposition pass in torch-spyre inductor

**Change Required:**
```python
# Current (incorrect):
torch.ops.prims.iota.default(n, start=0, step=1, dtype=torch.int64, 
                             device=device(type='cpu'), ...)

# Should be:
torch.ops.prims.iota.default(n, start=0, step=1, dtype=torch.int64, 
                             device=original_device, ...)
```

**Implementation:**
- Modify the decomposition logic to preserve the original device specification
- Ensure device information flows through all decomposition passes
- Add tests to verify device preservation

#### 2. Add Spyre Support for prims.iota (Priority: HIGH)

**Approach:**
- Implement `prims.iota` lowering for Spyre backend
- Handle int64 dtype appropriately (convert if needed)
- Ensure generated code uses Spyre device allocation

#### 3. Disable Decomposition for Small Tensors (Priority: MEDIUM)

**Rationale:** Passing tests show that direct `aten.arange` calls work correctly

**Implementation:**
- Add heuristic to skip decomposition for torch.arange
- Use direct aten.arange call when target device is Spyre
- Benchmark to ensure no performance regression

### Workarounds

#### Option 1: Type Conversion (Short-term)

```python
# Convert to float16 after generation
indices = torch.arange(height, device='spyre').to(torch.float16)
```

**Pros:** Simple, works immediately  
**Cons:** May lose precision for large indices, not suitable for all use cases

#### Option 2: CPU Fallback with Transfer (Short-term)

```python
# Generate on CPU, transfer to Spyre
indices = torch.arange(height, device='cpu').to('spyre')
```

**Pros:** Maintains int64 precision  
**Cons:** Performance overhead from device transfer

#### Option 3: Pre-compute Static Indices (Model-specific)

For models with fixed input sizes:
```python
# Pre-compute and cache position indices
position_ids = torch.arange(max_height, device='spyre')
```

**Pros:** Zero runtime overhead  
**Cons:** Only works for fixed-size inputs

### Long-term Solutions

1. **Comprehensive int64 Support:** Implement full int64 support on Spyre for common operations
2. **Smart Device Placement:** Develop heuristics to automatically choose optimal device for each operation
3. **Hybrid Execution:** Allow seamless CPU-Spyre execution for operations with limited support

---

## Testing Recommendations

### Regression Tests

Add test cases covering:

1. **Various Tensor Sizes:**
   ```python
   torch.arange(1, device='spyre')    # Small
   torch.arange(28, device='spyre')   # Medium
   torch.arange(1024, device='spyre') # Large
   ```

2. **Different Data Types:**
   ```python
   torch.arange(10, dtype=torch.int32, device='spyre')
   torch.arange(10, dtype=torch.int64, device='spyre')
   torch.arange(10, dtype=torch.float16, device='spyre')
   ```

3. **With Start/Step Parameters:**
   ```python
   torch.arange(0, 100, 2, device='spyre')
   torch.arange(10, 0, -1, device='spyre')
   ```

### Validation Tests

After implementing fixes:

1. **Device Verification:** Ensure output is on correct device
2. **Correctness:** Compare results with CPU reference
3. **Performance:** Benchmark against CPU execution
4. **Memory:** Verify no memory leaks or excessive allocation

---

## Related GitHub Issues

### Existing Issues

1. **[Issue #1153](https://github.com/torch-spyre/torch-spyre/issues/1153)** - Data type conversion
   - Relevance: int64 support limitations
   - Status: Open

2. **[Issue #1334](https://github.com/torch-spyre/torch-spyre/issues/1334)** - Scalar tensors
   - Relevance: Edge cases in tensor operations
   - Status: Open

### Recommended New Issue

**Title:** "torch.arange decomposition to prims.iota forces CPU device placement"

**Description:**
```markdown
## Issue
torch.arange operations are being decomposed to prims.iota with hardcoded 
device='cpu', causing device mismatch failures when target device is Spyre.

## Reproduction
See test: test_model_ops_db_torch_arange_1_spyre__Mistral-Small-3_2-24B-Instruct-2506

## Expected Behavior
Output tensor should be on Spyre device

## Actual Behavior
Output tensor is on CPU device

## Root Cause
Decomposition pass doesn't propagate device information to prims.iota

## Proposed Fix
1. Preserve device in decomposition
2. Implement prims.iota lowering for Spyre
3. Add device propagation tests
```

---

## Appendix: Compiler Artifacts

### A. Complete FX Graph (Failed Case)

```python
class <lambda>(torch.nn.Module):
    def forward(self):
        # File: site-packages/torch/testing/_internal/opinfo/core.py:1186
        iota: "i64[28]" = torch.ops.prims.iota.default(
            28, 
            start=0, 
            step=1, 
            dtype=torch.int64, 
            device=device(type='cpu'), 
            requires_grad=False
        )
        return (iota,)
```

### B. Complete IR (Failed Case)

```
op0: SchedulerNode(ComputedBuffer)
op0.writes = [MemoryDep('buf0', c0, {c0: 28})]
op0.unmet_dependencies = []
op0.met_dependencies = []
op0.outputs = [
    buf0: ComputedBuffer
    buf0.layout = FixedLayout('cpu', torch.int64, size=[28], stride=[1])
    buf0.users = [NodeUser(node=OUTPUT, can_inplace=False, is_weak=False)]
]
op0.group.device = cpu
op0.group.iteration = ((28,), ())
op0.sizes = ([28], [])
buf0_layout = FixedLayout('cpu', torch.int64, size=[28], stride=[1])
class op0_loop_body:
    var_ranges = {p0: 28}
    index0 = p0
    def body(self, ops):
        get_index = self.get_index('index0')
        index_expr = ops.index_expr(get_index, torch.int64)
        get_index_1 = self.get_index('index0')
        store = ops.store('buf0', get_index_1, index_expr, None)
        return store
```

### C. Generated C++ Kernel (Failed Case)

```cpp
extern "C" void kernel(int64_t* out_ptr0)
{
    {
        for(int64_t x0=static_cast<int64_t>(0L); x0<static_cast<int64_t>(28L); x0+=static_cast<int64_t>(16L))
        {
            {
                if(C10_LIKELY(x0 >= static_cast<int64_t>(0) && x0 < static_cast<int64_t>(16L)))
                {
                    auto tmp0 = x0;
                    auto tmp1 = c10::convert<int64_t>(tmp0);
                    auto tmp2 = at::vec::VectorizedN<int64_t,2>::arange(tmp1, 1);
                    tmp2.store(out_ptr0 + static_cast<int64_t>(x0), static_cast<int64_t>(16));
                }
                if(C10_UNLIKELY(x0 >= static_cast<int64_t>(16L) && x0 < static_cast<int64_t>(28L)))
                {
                    auto tmp0 = x0;
                    auto tmp1 = c10::convert<int64_t>(tmp0);
                    auto tmp2 = at::vec::VectorizedN<int64_t,2>::arange(tmp1, 1);
                    tmp2.store(out_ptr0 + static_cast<int64_t>(x0), static_cast<int64_t>(12L));
                }
            }
        }
    }
}
```

**Note:** This is a CPU kernel using vectorized operations, confirming CPU execution.

---

## Conclusion

The `torch.arange` test failure for Mistral-Small-3.2-24B-Instruct-2506 is caused by incorrect device placement during the decomposition of `torch.arange` to `prims.iota`. The decomposition logic hardcodes `device='cpu'`, overriding the intended Spyre device specification.

**Priority Actions:**
1. Fix decomposition to preserve device information
2. Implement Spyre backend support for prims.iota
3. Add comprehensive tests for device propagation

**Success Criteria:**
- All torch.arange tests pass with correct device placement
- No performance regression compared to current passing tests
- Proper handling of int64 dtype on Spyre or graceful fallback

---

**Report Generated:** 2026-04-07  
**Analyst:** Bob (AI Software Engineer)  
**Analysis Tool:** Claude Code (Windsurf IDE)
</details>

--- a/tests/models/runner.py
+++ b/tests/models/runner.py
@@ -359,7 +359,29 @@ def run_test(
 
     ref_out_cpu = to_device(ref_out, torch.device("cpu"))
 
-    # Check if the operation explicitly specifies CPU device
+    # Check if the operation specifies CPU device
+    FACTORY_METHODS = {
+        "torch.tensor",
+        "torch.from_numpy",
+        "torch.from_dlpack",
+        "torch.arange",
+        "torch.range",
+        "torch.linspace",
+        "torch.logspace",
+        "torch.zeros",
+        "torch.ones",
+        "torch.full",
+        "torch.eye",
+        "torch.rand",
+        "torch,randn",
+        "torch.randint",
+        "torch.randperm",
+        "torch.normal",
+        "torch.empty",
+        "torch.empty_stride",
+        "torch.sparse_coo_tensor",
+        "torch.complex",
+    }
     is_cpu_operation = False
     if op_name == "torch.to":
         # Check if device argument for torch.to is "cpu"
@@ -370,8 +392,12 @@ def run_test(
         ):
             is_cpu_operation = True
     else:
+        device_str = str(test_sample.kwargs.get("device", ""))
+        # default device for Pytorch factory method is "cpu"
+        if device_str == "" and op_name in FACTORY_METHODS:
+            is_cpu_operation = True
         # Check if device kwarg is "cpu"
-        if str(test_sample.kwargs.get("device", "")) == "cpu":
+        elif device_str == "cpu":
             is_cpu_operation = True
 
     # Check if the output tensor is on expected device

